hnswlib/hnswalg.h

```diff
@@ -800,6 +812,48 @@ class HierarchicalNSW : public AlgorithmInterface<dist_t, filter_func_t> {
     }
 
 
+    /**
+    * Adds point and replaces previously deleted point if any, updating it with new point
+    *
+    * If deleted point was replaced returns its label, else returns label of added point
+    */
+    labeltype insertPoint(const void* data_point, labeltype label) {
+        if (!replace_deleted) {
+            throw std::runtime_error("Don't use insertPoint when replacement of deleted elements is disabled");
+        }
+
+        std::unique_lock <std::mutex> tmp_del_el_lock(deleted_elements_lock);
+        bool is_empty = deleted_elements.empty();
+        tmp_del_el_lock.unlock();
+
+        if (is_empty) {
+            addPoint(data_point, label);
+            return label;
+        }
+        else {
+            tmp_del_el_lock.lock();
```

Comment:
It will crash here with parallel additions as another thread can access it between the locks.